### PR TITLE
Ladders take left/right clicks to go up or down (+ extra balance and QOL)

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
See the title. You now left click to climb up and right click to climb down a ladder. A delay of 1 second has also been added, since otherwise it'd take only one click to immediately move vertically and would be much more spammable.

Ghosts still use the old radials, because their right clicks are bound to the default byond popup menu.

## Why It's Good For The Game
QOL, bringing down the clicks necessary to climb a ladder up and down from 2 to 1. It has been tested on a local server.

## Changelog

:cl:
qol: Ladders now take left/right clicks to go up or down. Radial menus for them should only show up when you reach the floor above or below but the ladder goes beyond it, so that you don't have to bother clicking a sprite covered by that of your mob.
qol: Added contextual screen info (shows up top-center of the screen when hovering your cursor over the object) and similar information when examining the ladder.
qol: Improved messages for using ladders. They also make use of balloon messages now.
balance: Climbing up or down a ladder now takes 1 second by default. Less spammable, barely more immersive.
/:cl:

### CODE REVIEW

**Comment by Colovorat** [comment]
> Seems like this might be a little unintuitive. Does it mention on examine?

it doesn't

**Comment by Ghommie** [comment]
> Contextual screentip is definitely necessary

Yeah, sounds like a good idea.

>though I would agree that I don't think this is very intuitive

I could argue the effort to understand how it works is also minimal and having a radial menu for something that won't have more than two options is not much better overall.
But yes, there will be some struggle as older player uncommit what they had previously memorized, but it'll be nothing compared to deconstruction and tool_acts of several structures currently being split between left and right clicks without much of a common denominator for each, or the change from intents to combat mode.

Also, I'm marking this as a draft until I've time to fix the code smells and implement the screentip suggestion, which probably won't be happening before the day after tomorrow.

**Comment by Mothblocks** [comment]
If you want a weird idea, I think the best way is actually:

1. Left click keeps radial (thus adding zero learning requirement for new players)
2. Double click goes up
3. Right click goes down

**Comment by Mothblocks** [comment]
I would definitely promote my concern, yeah. I really don't want to add an extra learning step here.

**Comment by MrDoomBringer** [comment]
@Mothblocks Right clicking is the secondary action. New players will figure this out incredibly quickly and apply it to ladders intuitively. 

Having the screen tip at the top is more than enough. New players will learn very quickly to look up at those if they don't know how to work something. If you're really insistent, have a balloon alert appear when someone tries to left-click up a top-floor ladder telling them "Hey you're at the top floor already, right-click to go down"

Having a double-click action and a radial menu AND a right click action is ridiculous.

**Comment by Mothblocks** [comment]
> If you're really insistent, have a balloon alert appear when someone tries to left-click up a top-floor ladder telling them "Hey you're at the top floor already, right-click to go down"

Okay I'm fine with that

**Comment by ATH1909** [comment]
how will this affect zombies, borgs, simplemobs, etc.?

**Comment by Ghommie** [comment]
> will this make romerol zombies unable to climb ladders? were they unable to climb ladders before this PR?

You can use ladders while holding items, so unless the abstract item that's zombie claws had been changed for the worse, they should be able to use ladders.

**Comment by Ghommie** [comment]
Ok, I've added screen context, examine info and balloon messages as suggested, and radial menus to help move through 3+ stories tall ladders. It should be now ready for an actual code review.

**Comment by Mothblocks** [comment]
> You can use ladders while holding items, so unless the abstract item that's zombie claws had been changed for the worse, they should be able to.

Can you double check this? Not for zombies but just for having items in your hands. `attack_hand_secondary` isn't called when you have something in your hand, for instance.

**Comment by Mothblocks** [comment]
Oh wait nevermind I see you're using `attack`

File: code/game/objects/structures/ladders.dm [hunk 1]
```diff
@@ -123,39 +177,77 @@
...
+	return TRUE
+
+/obj/structure/ladder/attack_animal(mob/user, list/modifiers)
+	var/is_right_clicking = LAZYACCESS(modifiers, RIGHT_CLICK)
+	use(user, going_up = !is_right_clicking)
+	return TRUE
 
-/obj/structure/ladder/attack_larva(mob/user)
-	return use(user)
+/obj/structure/ladder/attack_slime(mob/user, list/modifiers)
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:207]: Can you replace these with some more core `attack_secondary`? I want to get rid of all the RIGHT_CLICK checks.

**Comment by Ghommie** [comment]
Ok, I've had to refactor unarmed attack code a little to fit your request in without filling the file with copypasta.
Due to this fact, until I test and confirm it's working as intended this is still a draft PR, but feel free to review the design.

**Comment by Ghommie** [comment]
Ok, it should be ready now.

File: code/_onclick/other_mobs.dm [hunk 2]
```diff
@@ -123,16 +124,27 @@
...
+/mob/living/proc/resolve_unarmed_attack(atom/attack_target, list/modifiers)
 	attack_target.attack_animal(src, modifiers)
 
+/mob/living/proc/resolve_right_click_attack(atom/target, list/modifiers)
+	return target.attack_animal_secondary(src, modifiers)
+
 /atom/proc/attack_animal(mob/user, list/modifiers)
 	SEND_SIGNAL(src, COMSIG_ATOM_ATTACK_ANIMAL, user)
 
+/atom/proc/attack_animal_secondary(mob/user, list/modifiers)
```
Comment (Mothblocks) [inline code/_onclick/other_mobs.dm:141]: Doc-comment

File: code/_onclick/other_mobs.dm [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
 	if (!LAZYACCESS(modifiers, RIGHT_CLICK))
 		return
-	var/secondary_result = target.attack_hand_secondary(src, modifiers)
+	var/secondary_result = resolve_right_click_attack(target, modifiers)
 
 	if (secondary_result == SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN || secondary_result == SECONDARY_ATTACK_CONTINUE_CHAIN)
 		return TRUE
 	else if (secondary_result != SECONDARY_ATTACK_CALL_NORMAL)
-		CRASH("attack_hand_secondary did not return a SECONDARY_ATTACK_* define.")
+		CRASH("resolve_right_click_attack did not return a SECONDARY_ATTACK_* define.")
```
Comment (Mothblocks) [inline code/_onclick/other_mobs.dm:10]:
```suggestion
		CRASH("resolve_right_click_attack (probably attack_hand_secondary) did not return a SECONDARY_ATTACK_* define.")
```

Be useful

File: code/_onclick/other_mobs.dm [hunk 4]
```diff
@@ -143,54 +155,69 @@
...
 	attack_target.attack_alien(src, modifiers)
 
+/mob/living/carbon/alien/resolve_right_click_attack(atom/target, list/modifiers)
+	return target.attack_alien_secondary(src, modifiers)
+
 /atom/proc/attack_alien(mob/living/carbon/alien/user, list/modifiers)
 	attack_paw(user, modifiers)
-	return
 
+/atom/proc/attack_alien_secondary(mob/living/carbon/alien/user, list/modifiers)
```
Comment (Mothblocks) [inline code/_onclick/other_mobs.dm:167]: Doc-comment

File: code/_onclick/other_mobs.dm [hunk 5]
```diff
@@ -143,54 +155,69 @@
...
+	attack_target.attack_larva(src, modifiers)
+
+/mob/living/carbon/alien/larva/resolve_right_click_attack(atom/target, list/modifiers)
+	return target.attack_larva_secondary(src, modifiers)
 
-/atom/proc/attack_larva(mob/user)
+/atom/proc/attack_larva(mob/user, list/modifiers)
 	return
 
+/atom/proc/attack_larva_secondary(mob/user, list/modifiers)
```
Comment (Mothblocks) [inline code/_onclick/other_mobs.dm:180]: Doc-comment

File: code/_onclick/other_mobs.dm [hunk 6]
```diff
@@ -143,54 +155,69 @@
...
-/atom/proc/attack_slime(mob/user)
+/mob/living/simple_animal/slime/resolve_right_click_attack(atom/target, list/modifiers)
+	if(isturf(target))
+		return ..()
+	return target.attack_slime_secondary(src, modifiers)
+
+/atom/proc/attack_slime(mob/user, list/modifiers)
 	return
 
+/atom/proc/attack_slime_secondary(mob/user, list/modifiers)
```
Comment (Mothblocks) [inline code/_onclick/other_mobs.dm:201]: Doc-comment

File: code/_onclick/other_mobs.dm [hunk 7]
```diff
@@ -143,54 +155,69 @@
...
 */
-/mob/living/simple_animal/drone/UnarmedAttack(atom/attack_target, proximity_flag, list/modifiers)
-	if(LIVING_UNARMED_ATTACK_BLOCKED(attack_target))
-		return
+
+/mob/living/simple_animal/drone/resolve_unarmed_attack(atom/attack_target, proximity_flag, list/modifiers)
 	attack_target.attack_drone(src, modifiers)
 
-/// Defaults to attack_hand or attack_hand_secondary. Override it when you don't want drones to do same stuff as humans.
+/mob/living/simple_animal/drone/resolve_right_click_attack(atom/target, list/modifiers)
```
Comment (Mothblocks) [inline code/_onclick/other_mobs.dm:211]: Doc-comment

File: code/game/objects/structures/ladders.dm [hunk 8]
```diff
@@ -70,50 +84,90 @@
 		visible_message(span_danger("[src] is torn to pieces by the gravitational pull!"))
 		qdel(src)
 
-/obj/structure/ladder/proc/travel(going_up, mob/user, is_ghost, obj/structure/ladder/ladder)
+/obj/structure/ladder/proc/use(mob/user, going_up = TRUE)
+	if(!in_range(src, user) || DOING_INTERACTION(user, DOAFTER_SOURCE_CLIMBING_LADDER))
+		return
+
+	if(!up && !down)
+		balloon_alert(user, span_warning("Doesn't lead anywhere!"))
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:92]:
```suggestion
		balloon_alert(user, span_warning("doesn't lead anywhere!"))
```

File: code/game/objects/structures/ladders.dm [hunk 9]
```diff
@@ -70,50 +84,90 @@
...
+		return
+	if(going_up ? !up : !down)
+		balloon_alert(user, span_warning("Can't go any further [going_up ? "up" : "down"]."))
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:95]:
```suggestion
		balloon_alert(user, span_warning("can't go any further [going_up ? "up" : "down"]!"))
```

File: code/game/objects/structures/ladders.dm [hunk 10]
```diff
@@ -70,50 +84,90 @@
...
+
+/obj/structure/ladder/proc/start_travelling(mob/user, going_up)
+	show_initial_fluff_message(user, going_up)
+	if(do_after(user, travel_time, target = src, interaction_key = DOAFTER_SOURCE_CLIMBING_LADDER))
+		travel(user, going_up)
+
+/// The message shown when the player starts climbing the ladder
+/obj/structure/ladder/proc/show_initial_fluff_message(mob/user, going_up)
+	var/up_down = going_up ? "up" : "down"
+	user.balloon_alert_to_viewers("climbing [up_down] [src]", "climbing [up_down] [src]")
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:111]:
```suggestion
	user.balloon_alert_to_viewers("climbing [up_down]...")
```

File: code/game/objects/structures/ladders.dm [hunk 11]
```diff
@@ -70,50 +84,90 @@
...
+
+/obj/structure/ladder/proc/travel(mob/user, going_up, is_ghost = FALSE)
+	var/obj/structure/ladder/ladder = going_up ? up : down
+	if(!ladder)
+		to_chat(user, span_warning("[src] doesn't seem to lead anywhere that way!"))
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:116]:
```suggestion
		balloon_alert(user, "there's nothing that way!")
```

File: code/game/objects/structures/ladders.dm [hunk 12]
```diff
@@ -70,50 +84,90 @@
...
+
+/// The messages shown after the player has finished climbing. Players can see this happen from either src or the destination so we've 2 POVs here
+/obj/structure/ladder/proc/show_final_fluff_message(mob/user, obj/structure/ladder/destination, going_up)
+	var/up_down = going_up ? "up" : "down"
+
+	//POV of players around the source
+	visible_message(span_notice("[user] climbs [up_down] [src]."))
+	//POV of players around the destination
+	user.visible_message(span_notice("[user] climbs [up_down] [destination] from [going_up ? "below" : "above"]."), span_notice("You climb down [src]."))
+	user.balloon_alert_to_viewers("climbed [up_down]", "climbed [up_down]")
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:142]:
```suggestion
	user.balloon_alert_to_viewers("climbed [up_down]")
```

File: code/game/objects/structures/ladders.dm [hunk 13]
```diff
@@ -70,50 +84,90 @@
...
+		ladder.show_options(user, is_ghost)
+
+/// The messages shown after the player has finished climbing. Players can see this happen from either src or the destination so we've 2 POVs here
+/obj/structure/ladder/proc/show_final_fluff_message(mob/user, obj/structure/ladder/destination, going_up)
+	var/up_down = going_up ? "up" : "down"
+
+	//POV of players around the source
+	visible_message(span_notice("[user] climbs [up_down] [src]."))
+	//POV of players around the destination
+	user.visible_message(span_notice("[user] climbs [up_down] [destination] from [going_up ? "below" : "above"]."), span_notice("You climb down [src]."))
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:141]:
```suggestion
```

Don't have both

File: code/game/objects/structures/ladders.dm [hunk 14]
```diff
@@ -123,39 +177,101 @@
...
-	else
-		user.visible_message(span_notice("[user] climbs down [src]."), span_notice("You climb down [src]."))
-
+///Ghosts use the byond default popup menu function on right click, so this is going to work a little differently for them.
+/obj/structure/ladder/proc/ghost_use(mob/user)
+	if (!up && !down)
+		balloon_alert(user, span_warning("It doesn't lead anywhere!"))
+		return
+	if(!up) //only goes down
+		travel(user, FALSE, TRUE)
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:270]: Named parameters

File: code/game/objects/structures/ladders.dm [hunk 15]
```diff
@@ -123,39 +177,101 @@
...
+	else if(!down) //only goes up
+		travel(user, TRUE, TRUE)
```
Comment (Mothblocks) [inline code/game/objects/structures/ladders.dm:272]: Named parameters

File: code/modules/awaymissions/mission_code/Cabin.dm [hunk 16]
```diff
@@ -114,11 +114,23 @@
...
+		context[SCREENTIP_CONTEXT_RMB] = "Warp down"
+	return CONTEXTUAL_SCREENTIP_SET
+
+/obj/structure/ladder/unbreakable/rune/show_initial_fluff_message(mob/user, going_up)
+	user.balloon_alert_to_viewers("activating [src]", "activating [src]")
+
+/obj/structure/ladder/unbreakable/rune/show_final_fluff_message(mob/user, going_up)
+	visible_message(span_notice("[user] activates [src] and teleports away."))
+	user.visible_message(span_notice("[user] warps in over [src]."), span_notice("You warp in over [src]."))
+	user.balloon_alert_to_viewers("warped in", "warped in")
```
Comment (Mothblocks) [inline code/modules/awaymissions/mission_code/Cabin.dm:130]:
```suggestion
	user.balloon_alert_to_viewers("warped in")
```

File: code/modules/awaymissions/mission_code/Cabin.dm [hunk 17]
```diff
@@ -114,11 +114,23 @@
...
+	if(down)
+		context[SCREENTIP_CONTEXT_RMB] = "Warp down"
+	return CONTEXTUAL_SCREENTIP_SET
+
+/obj/structure/ladder/unbreakable/rune/show_initial_fluff_message(mob/user, going_up)
+	user.balloon_alert_to_viewers("activating [src]", "activating [src]")
+
+/obj/structure/ladder/unbreakable/rune/show_final_fluff_message(mob/user, going_up)
+	visible_message(span_notice("[user] activates [src] and teleports away."))
+	user.visible_message(span_notice("[user] warps in over [src]."), span_notice("You warp in over [src]."))
```
Comment (Mothblocks) [inline code/modules/awaymissions/mission_code/Cabin.dm:129]:
```suggestion
```

Don't have both

File: code/modules/awaymissions/mission_code/Cabin.dm [hunk 18]
```diff
@@ -114,11 +114,23 @@
...
-	if(is_ghost || !IS_WIZARD(user))
+/obj/structure/ladder/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	if(up)
+		context[SCREENTIP_CONTEXT_LMB] = "Warp up"
+	if(down)
+		context[SCREENTIP_CONTEXT_RMB] = "Warp down"
+	return CONTEXTUAL_SCREENTIP_SET
+
+/obj/structure/ladder/unbreakable/rune/show_initial_fluff_message(mob/user, going_up)
+	user.balloon_alert_to_viewers("activating [src]", "activating [src]")
```
Comment (Mothblocks) [inline code/modules/awaymissions/mission_code/Cabin.dm:125]:
```suggestion
	user.balloon_alert_to_viewers("activating...")
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Mothblocks** [comment]
Closer, thanks

**Comment by Ghommie** [comment]
Done. Can you add a refactor tag to it now by the by? Your sensible request for more attack_*_secondary procs had me refactoring the click sequence a bit to minimize the bloat.

**Comment by Fikou** [comment]
moth is on holiday so ill merge this

**Comment by RikuTheKiller** [comment]
I've hated this for ages now and I finally feel like making a PR to remove the delay. (Later ofc, procrastination.)

Stairs don't have a delay either and moving around generally shouldn't have delays at all except for doors.

It's such a fucking sequence break to go up a ladder only for it to glitch out, send me back down only for me to have to wait another few useless seconds because "but muh immersion..." PRs like this exist.

**Comment by Mothblocks** [comment]
I'm fine removing the delay FWIW

**Comment by Ghommie** [comment]
Feel free to remove the delay if you want to. It wasn't the main scope of this PR.

**Comment by RikuTheKiller** [comment]
Will do.

**Comment by Mothblocks** [comment]
Sounds like you should fix those bugs instead? The changelog provides a good motivation for the radial menu existing

**Comment by RikuTheKiller** [comment]
> Sounds like you should fix those bugs instead? The changelog provides a good motivation for the radial menu existing

That's true.

**Pull request merged**